lls/main.c

```diff
@@ -719,6 +867,58 @@ register_nd_acl_rules(struct gatekeeper_if *iface)
 	return 0;
 }
 
+static void
+fill_ping6_rule(struct ipv6_acl_rule *rule, struct in6_addr *addr)
+{
+	uint32_t *ptr32 = (uint32_t *)addr;
+	int i;
+
+	rule->data.category_mask = 0x1;
+	rule->data.priority = 1;
+	/* Userdata is filled in in register_ipv6_acl(). */
+
+	rule->field[PROTO_FIELD_IPV6].value.u8 = IPPROTO_ICMPV6;
+	rule->field[PROTO_FIELD_IPV6].mask_range.u8 = 0xFF;
+
+	for (i = DST1_FIELD_IPV6; i <= DST4_FIELD_IPV6; i++) {
+		rule->field[i].value.u32 = rte_be_to_cpu_32(*ptr32);
+		rule->field[i].mask_range.u32 = 32;
+		ptr32++;
+	}
+
+	rule->field[TYPE_FIELD_ICMPV6].value.u32 =
+		(ICMPV6_ECHO_REQUEST_TYPE << 24) & 0xFFFF0000;
+	rule->field[TYPE_FIELD_ICMPV6].mask_range.u32 = 0xFFFF0000;
+}
+
+static int
+register_ping6_acl_rules(struct gatekeeper_if *iface)
+{
+	struct ipv6_acl_rule ipv6_rules[NUM_ACL_PING_RULES];
+	int ret;
+
+	memset(&ipv6_rules, 0, sizeof(ipv6_rules));
+
+	/*
+	 * According to RFC 4443 section 4.1, the Destination Address of an
+	 * ICMPv6 Echo Request message can be any legal IPv6 address.
+	 */
+	fill_ping6_rule(&ipv6_rules[0], &iface->ip6_addr);
+	fill_ping6_rule(&ipv6_rules[1], &iface->ll_ip6_addr);
+	fill_ping6_rule(&ipv6_rules[2], &iface->ip6_mc_addr);
+	fill_ping6_rule(&ipv6_rules[3], &iface->ll_ip6_mc_addr);
+
+	ret = register_ipv6_acl(ipv6_rules, NUM_ACL_PING_RULES,
+		reply_ping6, match_ping6, iface);
```

Comment:
We cannot pass `reply_ping6()` here because it would cause a number of problems since each GK/GT instance calls concurrently this function. See `submit_arp()` or `submit_nd()` for an example of how to do it.

This comment also applies to patch **lls: respond to ICMP pings**.